File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/PerfilService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.model.Perfil;
import br.com.portalproposta.backend.repository.PerfilRepository;
import br.com.portalproposta.backend.repository.UsuarioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class PerfilService {

    private final PerfilRepository perfilRepository;
    private final UsuarioRepository usuarioRepository;

    @Autowired
    public PerfilService(PerfilRepository perfilRepository, UsuarioRepository usuarioRepository) {
        this.perfilRepository = perfilRepository;
        this.usuarioRepository = usuarioRepository;
    }

    public List<Perfil> list() {
        return perfilRepository.findAllByOrderByNomeAsc();
    }

    public Perfil getById(String id) {
        return perfilRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("perfil de acesso não encontrado"));
    }

    @Transactional
    public Perfil create(Perfil perfil) {
        if (perfil.getNome() == null || perfil.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do perfil é obrigatório");
        }

        perfilRepository.findByNome(perfil.getNome()).ifPresent(existing -> {
            throw new AppBadRequestException("já existe um perfil com este nome");
        });

        perfil.setId(UUID.randomUUID().toString());
        perfil.setCreatedAt(LocalDateTime.now());
        perfil.setUpdatedAt(LocalDateTime.now());

        return perfilRepository.save(perfil);
    }

    @Transactional
    public Perfil update(String id, Perfil input) {
        Perfil perfil = getById(id);

        if (input.getNome() == null || input.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do perfil é obrigatório");
        }

        if (!input.getNome().equals(perfil.getNome())) {
            perfilRepository.findByNome(input.getNome()).ifPresent(existing -> {
                if (!existing.getId().equals(perfil.getId())) {
                    throw new AppBadRequestException("já existe um perfil com este nome");
                }
            });
        }

        if (!perfil.isSistema()) {
            perfil.setNome(input.getNome());
        }

        perfil.setDescricao(input.getDescricao());
        perfil.setPermissoes(input.getPermissoes());
        perfil.setUpdatedAt(LocalDateTime.now());

        return perfilRepository.save(perfil);
    }

    @Transactional
    public void delete(String id) {
        Perfil perfil = getById(id);

        if (perfil.isSistema()) {
            throw new AppBadRequestException("não é permitido excluir perfis do sistema");
        }

        if (usuarioRepository.existsByPerfilId(id)) {
            // Count matching users
            long count = usuarioRepository.findAll().stream().filter(u -> u.getPerfilId().equals(id)).count();
            throw new AppBadRequestException(String
                    .format("não é possível excluir este perfil pois existem %d usuário(s) associado(s) a ele", count));
        }

        perfilRepository.delete(perfil);
    }
}

File: backend-net10/Services/PerfilService.cs
using Microsoft.EntityFrameworkCore;
using backend_net10.Data;
using backend_net10.Models;

namespace backend_net10.Services;

public class PerfilService
{
    private readonly AppDbContext _context;

    public PerfilService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<List<Perfil>> ListAsync()
    {
        return await _context.Perfis.OrderBy(p => p.Nome).ToListAsync();
    }

    public async Task<Perfil> GetByIdAsync(string id)
    {
        var perfil = await _context.Perfis.FirstOrDefaultAsync(p => p.Id == id);
        if (perfil == null)
        {
            throw new AppNotFoundException("perfil de acesso não encontrado");
        }
        return perfil;
    }

    public async Task<Perfil> CreateAsync(Perfil perfil)
    {
        if (string.IsNullOrWhiteSpace(perfil.Nome))
        {
            throw new AppBadRequestException("nome do perfil é obrigatório");
        }

        var existing = await _context.Perfis.FirstOrDefaultAsync(p => p.Nome == perfil.Nome);
        if (existing != null)
        {
            throw new AppBadRequestException("já existe um perfil com este nome");
        }

        perfil.Id = Guid.NewGuid().ToString();
        perfil.CreatedAt = DateTime.Now;
        perfil.UpdatedAt = DateTime.Now;

        _context.Perfis.Add(perfil);
        await _context.SaveChangesAsync();

        return perfil;
    }

    public async Task<Perfil> UpdateAsync(string id, Perfil input)
    {
        var perfil = await _context.Perfis.FirstOrDefaultAsync(p => p.Id == id);
        if (perfil == null)
        {
            throw new AppNotFoundException("perfil de acesso não encontrado");
        }

        if (string.IsNullOrWhiteSpace(input.Nome))
        {
            throw new AppBadRequestException("nome do perfil é obrigatório");
        }

        if (input.Nome != perfil.Nome)
        {
            var existing = await _context.Perfis.FirstOrDefaultAsync(p => p.Nome == input.Nome);
            if (existing != null && existing.Id != perfil.Id)
            {
                throw new AppBadRequestException("já existe um perfil com este nome");
            }
        }

        if (!perfil.IsSistema)
        {
            perfil.Nome = input.Nome;
        }

        perfil.Descricao = input.Descricao;
        perfil.Permissoes = input.Permissoes;
        perfil.UpdatedAt = DateTime.Now;

        await _context.SaveChangesAsync();
        return perfil;
    }

    public async Task DeleteAsync(string id)
    {
        var perfil = await _context.Perfis.FirstOrDefaultAsync(p => p.Id == id);
        if (perfil == null)
        {
            throw new AppNotFoundException("perfil de acesso não encontrado");
        }

        if (perfil.IsSistema)
        {
            throw new AppBadRequestException("não é permitido excluir perfis do sistema");
        }

        var count = await _context.Usuarios.CountAsync(u => u.PerfilId == id);
        if (count > 0)
        {
            throw new AppBadRequestException($"não é possível excluir este perfil pois existem {count} usuário(s) associado(s) a ele");
        }

        _context.Perfis.Remove(perfil);
        await _context.SaveChangesAsync();
    }
}
